components/fxa-client/src/jwcrypto.rs

```diff
@@ -0,0 +1,265 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+/// Theorically, everything done in this file could and should be done in a JWT library.
+/// However, none of the existing rust JWT libraries can handle ECDH-ES encryption, and API choices
+/// made by their authors make it difficult to add this feature.
+/// In the past, we chose cjose to do that job, but it added three C dependencies to build and link
+/// against: jansson, openssl and cjose itself.
+use crate::error::*;
+use rc_crypto::{
+    aead, agreement,
+    agreement::{Ephemeral, InputKeyMaterial, KeyPair},
+    digest, rand,
+};
+use serde_derive::{Deserialize, Serialize};
+
+pub enum EncryptionParameters {
+    // ECDH-ES in Direct Key Agreement mode.
+    #[allow(non_camel_case_types)]
+    ECDH_ES {
+        enc: EncryptionAlgorithm,
+        peer_jwk: Jwk,
+        local_key_pair: KeyPair<Ephemeral>,
+    },
+}
+
+pub enum DecryptionParameters {
+    // ECDH-ES in Direct Key Agreement mode.
+    #[allow(non_camel_case_types)]
+    ECDH_ES { local_key_pair: KeyPair<Ephemeral> },
+}
+
+#[derive(Serialize, Deserialize, PartialEq)]
+enum Algorithm {
+    #[serde(rename = "ECDH-ES")]
+    #[allow(non_camel_case_types)]
+    ECDH_ES,
+}
+#[derive(Serialize, Deserialize, PartialEq)]
+pub enum EncryptionAlgorithm {
+    A256GCM,
+}
+
+impl EncryptionAlgorithm {
+    fn algorithm_id(&self) -> &'static str {
+        match self {
+            Self::A256GCM => "A256GCM",
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize)]
+struct JweHeader {
+    pub alg: Algorithm,
+    pub enc: EncryptionAlgorithm,
+    pub epk: Option<Jwk>,
+    pub apu: Option<String>,
+    pub apv: Option<String>,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct Jwk {
+    pub crv: String,
+    pub kty: String,
+    pub x: String,
+    pub y: String,
+}
+
+/// Encrypt and serialize data in the JWE compact form.
+pub fn encrypt_to_jwe(data: &[u8], encryption_params: EncryptionParameters) -> Result<String> {
+    let (encrypted_key, protected_header, iv, ciphertext, auth_tag) = match encryption_params {
+        EncryptionParameters::ECDH_ES {
+            enc,
+            local_key_pair,
+            peer_jwk,
+        } => {
+            let local_public_key = extract_pub_key_jwk(&local_key_pair)?;
```

Comment:
Rather than taking `epk` as an input parameter here, I think this method should generate a fresh keypair internally. This is both safer (since we can ensure that the key really is only used a single time) and simpler for the caller (since it doesn't have to figure out how to generate such a key).

Reply:
This would also allow this https://github.com/mozilla/application-services/blob/445c58fe8fb2cb612fe4a3b50fcfdc89eb1a024b/components/fxa-client/src/oauth.rs#L234-L235 to no longer have to depend on creating a "ScopedKeysFlow" as well!